### ISSUE TITLE
bmerge errors on integer64 to double join

### ISSUE THREAD

**Issue by MichaelChirico** [Rdatatable/data.table#4167]
Just came across this... haven't had time to minimize the example but here goes:

```
library(DBI)
library(RSQLite)
library(data.table)
# download and unzip file from here:
#   https://ankiweb.net/shared/info/867291675
card_db = dbConnect(SQLite(),  'collection.anki2')
cards = setDT(dbGetQuery(card_db, 'select * from cards'))
cards[id == 1419644220828]
# Error in bmerge(i, x, leftcols, rightcols, roll, rollends, nomatch, mult,  : 
#   Incompatible join types: x.id is type integer64 but i.id is type double and contains fractions
```

**Comment by jangorecki** [comment]
```r
cards=data.table(id=bit64::as.integer64(1419644220828))
cards[id == 1419644220828]
# Error in bmerge(i, x, leftcols, rightcols, roll, rollends, nomatch, mult,  : 
#   Incompatible join types: x.id is type integer64 but i.id is type double and contains fractions
```

**Comment by MichaelChirico** [comment]
Forgot to add `Low`, filed because:

> and contains fractions

is not right

**Comment by vspinu** [comment]
I think there is no reliable way to check if a double is an "integer" or not. So I think it's best in this [line](https://github.com/Rdatatable/data.table/blob/57361d6974f81f976a7fb1ce28ef003592817b78/R/bmerge.R#L93) to just convert to integer64 and rely on bit64 to actually do the check. 

Or at least add an option to disable the check. Or maybe just add a warning instead of an error.

**Comment by jangorecki** [comment]
I think we should promote integer64 to double type instead.
Logical hierarchy would be like this
integer < integer64 < numeric
The actual storage precision will not match to that hierarchy bc:
2^31 < 2^63 < 2^51
But this is not our problem, what we should care about is logical hierarchy

**Comment by MichaelChirico** [comment]
Just found this bug again after I filed the duplicate #6625.

> I think we should promote integer64 to double type instead.

I'm not sure I follow. I think the approach taken in #6626 is OK.

Fundamentally, there is no hierarchical ordering of `integer64` and `double`. There are many integers that can be represented as `integer64`, but not as `double`, and vice versa. That contrasts to `integer32` which is a strict subset of `double`.

For example, `36028797018963972` (`2^55 + 4`) is perfectly represented in `integer64`, but resolves as `36028797018963968` in `double`, whereas `1180591620717411303424` (`2^70`) is perfectly represented in `double`, but is outside the `integer64` limits. (see https://numeral-systems.com/ieee-754-converter/)

**Comment by MichaelChirico** [comment]
Trying to write out a bit more carefully why I think the approach taken in #6626 is good.

 - If we want to attempt the join at all, we need a common storage type. Either `integer64` or `double` should be chosen.
   + 128-bit storage might be better but is not really that widespread: https://en.wikipedia.org/wiki/Quadruple-precision_floating-point_format "quadruple precision may be specified by the [long double](https://en.wikipedia.org/wiki/Long_double) type, but this is not required by the language (which only requires long double to be at least as precise as double), nor is it common"
   + `character` is an available common super-type, but coercing to character to join is definitely slow. I'd rather error and ask the user to coerce on their own than pursue this.
 - We should not try to convert `integer64`➡️`double` because there's not easy way to ensure it's possible IINM. `(double)i64 == i64` won't really work, IIUC it'd have to be `(long double)i64` _and_ be assured `long double` is actually quadruple-precision (which, see above). Other ways to check would require lots of handwritten arithmetic that will be hard to assure is platform- and implementation-robust.
 - By contrast, `isRealReallyInt64()` looks reasonable to me as a way to ensure converting `double`➡️`integer64` WAI. To be assured a given `double` has representation as `integer64`, it's enough to know that (1) the value is within `bit64::lim.integer64()` range and (2) it has no fractional part.
 - The downside is there are some joins that might be possible which we don't perform, e.g. joining `2^100` to `as.integer64(1L)` in principal is the same as joining to `as.integer64(2L)` -- it's just like any other miss to be handled by whether we're doing `INNER` or `OUTER` join.
   + This is fine, the user should just do the coercion themselves if they want to proceed.
